kimchi/src/circuits/cairo/runner/program.rs

```diff
@@ -0,0 +1,466 @@
+//! This module represents a Cairo execution step
+//! It defines the execution logic of Cairo instructions
+
+use crate::circuits::cairo::runner::definitions::*;
+use crate::circuits::cairo::runner::memory::CairoMemory;
+use crate::circuits::cairo::runner::word::CairoWord;
+use ark_ff::PrimeField;
+
+/// A structure to store program counter, allocation pointer and frame pointer
+#[derive(Clone, Copy)]
+pub struct CairoPointers<F: PrimeField> {
+    /// Program counter: points to address in memory
+    pub pc: F,
+    /// Allocation pointer: points to first free space in memory
+    pub ap: F,
+    /// Frame pointer: points to the beginning of the stack in memory (for arguments)
+    pub fp: F,
+}
+
+impl<F: PrimeField> CairoPointers<F> {
+    /// Creates a new triple of pointers
+    pub fn new(pc: F, ap: F, fp: F) -> CairoPointers<F> {
+        CairoPointers { pc, ap, fp }
+    }
+}
+
+/// A structure to store auxiliary variables throughout computation
+pub struct CairoVariables<F: PrimeField> {
+    /// Destination
+    dst: Option<F>,
+    /// First operand
+    op0: Option<F>,
+    /// Second operand
+    op1: Option<F>,
+    /// Result
+    res: Option<F>,
+    /// Destination address
+    dst_addr: F,
+    /// First operand address
+    op0_addr: F,
+    /// Second operand address
+    op1_addr: F,
+    /// Size of the instruction
+    size: F,
+}
+
+impl<F: PrimeField> CairoVariables<F> {
+    /// This function creates an instance of a default CairoVariables struct
+    pub fn new() -> CairoVariables<F> {
+        CairoVariables {
+            dst: None,
+            op0: None,
+            op1: None,
+            res: None,
+            dst_addr: F::zero(),
+            op0_addr: F::zero(),
+            op1_addr: F::zero(),
+            size: F::zero(),
+        }
+    }
+}
+impl<F: PrimeField> Default for CairoVariables<F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A data structure to store a current step of Cairo computation
+pub struct CairoStep<'a, F: PrimeField> {
+    // current step of computation
+    //step: u64,
+    /// current word of the program
+    pub mem: &'a mut CairoMemory<F>,
```

Comment:
I chose a mutable reference with a given lifetime because I didn't want to copy or clone the whole structure of `CairoMemory` every time it was being called in a different function. In particular, since each step is given by the status of the memory at that time and a series of variables, and then the next step takes the status of memory and builds on top of it, I thought the program could just be "passing" the same object for every step. So mainly, the reason is potential efficiency and saving some space.